pydantic/_internal/_generate_schema.py

```diff
@@ -533,36 +529,35 @@ def _apply_discriminator_to_union(
                 discriminator,
             )
         except _discriminated_union.MissingDefinitionForUnionRef:
-            # defer until defs are resolved
-            _discriminated_union.set_discriminator_in_metadata(
-                schema,
-                discriminator,
-            )
+            # defer until defs are resolved.
+            _discriminated_union.set_discriminator_in_metadata(schema, discriminator)
+            self.defs.used_deferred_discriminators = True
             return schema
 
-    class CollectedInvalid(Exception):
-        pass
-
-    def clean_schema(self, schema: CoreSchema) -> CoreSchema:
-        schema = self.collect_definitions(schema)
-        schema = simplify_schema_references(schema)
-        if collect_invalid_schemas(schema):
-            raise self.CollectedInvalid()
-        schema = _discriminated_union.apply_discriminators(schema)
+    def clean_schema(self, schema: CoreSchema, *, deep_copy: bool) -> CoreSchema:
+        """Cleans the schema contents by substituting definition-refs and deferred discriminators. Given schema must be
+        the one just given by generate_schema. Does in place substituting to the schema by mutating it for performance
+        reasons. Use deep_copy to avoid mutating the original schema.
+
+        Example for substituting definition-refs:
+        Input:
+        {"type": "definition-ref", "schema_ref": "Xyz"}
+        Definitions:
+        {"Xyz": {"type": "string", "ref": "Xyz"}}
+        Output:
+        {"type": "string", "ref": "Xyz"}
```

Comment:
Maybe this is just an oversimplification, but seems confusing that something of type `"string"` would have a ref...

Reply:
It happens with type alias usage. Eg with `TypeAdapter(TypeAliasType('A', int))` this happens exactly :D